### DISCUSSION TITLE
How to free Free OptiX IAS?

### DISCUSSION THREAD

**Discussion by h-OUS-e** [mitsuba-renderer/mitsuba3#627] [Q&A] [answered]
Hello,

I am trying to create a fastAPI app with Mitsuba. The algorithm goes something like this:

```ruby
1. Read a mesh surface
2. Add mesh surface to a scene dictionary
3. mi.dict_load() the scene dictionary
4. ray trace the scene inside a sample() function with a custom mi.loop()
```

On the first try, the ray tracing works and a scene is built with Optix on the first mi.Loop():
```
2023-04-02 01:19:42 INFO  main  [Scene] Building scene in OptiX ..
2023-04-02 01:19:42 DEBUG main  [scene_optix.inl:73] Initialize Optix configuration (index=2)..
2023-04-02 01:19:42 INFO  main  [Scene] OptiX ready. (took 93ms)
Starting loop
Loop Done
```
But once I get a new mesh surface, and make a new scene to add this new mesh, running the `sample()` function returns:
```
jit_eval(): more than one OptiX shader binding table was used within a single kernel, which is not supported. Please split your kernel into smaller parts (e.g. using `dr::eval()`). Disabling the ray tracing operation to avoid potential undefined behavior.
```

This message doesn't show when I use `mi.render()`. Looking at the logger, running `mi.render()` prints `[Scene] Free OptiX IAS..`. I was wondering if I can also free up the Optix IAS inside python with dr.jit to run this custom `mi.loop()` on the server with updated scenes?

**Comment by njroussel** [comment, marked as answer]
Hi @h-OUS-e 

We don't expose a way to manually delete the OptiX acceleration structures, for good reasons: you might have some variables that depend on it that have not yet been evaluated.
They delete themselves when the scene can be de-allocated. This usually happens as soon as it is no longer referenceable and isn't a dependency of any other variables.

**Reply by h-OUS-e** [reply]
`del scene`, or set `scene=None` didn't work for me, but shouldn't that have made the scene no longer reference-able? What is the right way to de-allocate the scene to free the Optix acceleration structures?

**Reply by njroussel** [reply]
Yes, but you might have defined a variable that references the scene in its computation before you did `del scene`.

A tiny example:
```python
a = mi.Float(1)
b = mi.Float(2)
c = a + b
del b # Will not delete b, because c needs it and c is not evaluated
print(c) # Evalutates c and frees b as it no longer referenced by any variable
```

**Reply by h-OUS-e** [reply]
I see, I had a sampler outside the web-socket function, and so it wasn't being reset, but now it works thank you! So much magic going on in the backend.
